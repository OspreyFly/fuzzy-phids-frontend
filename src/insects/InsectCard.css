/* InsectCard.css */
.card {
  border: 1px solid #ccc;
  border-radius: 10px;
  box-shadow: 0 12px 24px rgba(0,91,65,0.5);
  margin: 15px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  background-color: #008170;
}

.card:hover{
  box-shadow: 0 24px 48px rgba(0,91,65,0.9);
}

.card-body {
  padding: 20px;
  color: #ffffff;
  text-align: center;
}

.card-title {
  margin-bottom: 10px;
  margin-top: 10px;
  font-size: 1.5em;

}

.card-title img {
  max-width: 100%; /* Ensure image doesn't exceed card width */
  height: auto;
  
}

h3 {
  font-size: 1.75em;
  margin-top: 0;
}
