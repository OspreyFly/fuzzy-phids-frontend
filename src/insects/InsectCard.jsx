import React, { useEffect, useState } from "react";
import "./InsectCard.css";

/** Show limited information about an insect
 *
 * Is rendered by InsectList to show a "card" for each insect.
 *
 * InsectList -> InsectCard
 */
const InsectCard = ({ id, species, price, url_image, cart, addToCart, removeFromCart }) => {
    const [isDisabled, setIsDisabled] = useState(false);

    useEffect(() => {
        // Check if the insectId exists in the cart
        const insectInCart = cart.some(item => item.id === id);

        // Set isDisabled to true if the insect is in the cart
        if (insectInCart) {
            setIsDisabled(true);
        } else {
            setIsDisabled(false);
        }
    }, [cart, id]); // Re-run the check whenever the cart or insectId changes

    const handleAddClick = () => {
        if (!isDisabled) {
            setIsDisabled(true);
            addToCart({ id, species, price, url_image });
        }
    }
    const handleRemoveClick = () => {
        removeFromCart(id);
    }

    return (
        <div className="InsectCard card">
            <div className="card-body">
                <h6 className="card-title">
                    {species}
                    {url_image && <img src={url_image} alt={species} className="float-right ml-5" />}
                </h6>
                <h3>${price}</h3>
                {!isDisabled && <button className="animated-plus-button" onClick={handleAddClick}>+</button>}
                {isDisabled && <button className="animated-minus-button" onClick={handleRemoveClick}>-</button>}
            </div>
        </div>
    );
}

export default InsectCard;
