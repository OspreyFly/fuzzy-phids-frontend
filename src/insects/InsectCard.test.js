import React from 'react';
import { render, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import InsectCard from './InsectCard';
import '@testing-library/jest-dom';


const cart = [{id:1, species: "a", price: 1}, {id:2, species: "b", price: 1}];

test('render matches snapshot', () => {
  const {asFragment} = render(<InsectCard cart={cart}/>);
  expect(asFragment()).toMatchSnapshot();
});

test('clicking the button will change the text from plus to minus', async () => {
  const user = userEvent.setup()

  render(<InsectCard cart={[]} addToCart={() => {}}/>);

  // Initially, the plus button should be visible
  expect(screen.getByRole('button', {name: '+'})).toBeInTheDocument();

  // Simulate adding an item to the cart
  await user.click(screen.getByRole('button', {name: "+"}))

  // After adding, the remove button should be visible
  expect(screen.getByRole('button', {name: '-'})).toBeInTheDocument();
})
