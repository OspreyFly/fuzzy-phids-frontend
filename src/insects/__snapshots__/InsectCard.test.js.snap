// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="InsectCard card"
  >
    <div
      class="card-body"
    >
      <h6
        class="card-title"
      />
      <h3>
        $
      </h3>
      <button
        class="animated-plus-button"
      >
        +
      </button>
    </div>
  </div>
</DocumentFragment>
`;
